### DISCUSSION TITLE
Two questions :) about PowerDown with AVR128DB28

### DISCUSSION THREAD

**Discussion by XFer012** [SpenceKonde/DxCore#250] [Q&A]
Hello,
I'm using DxCore v1.4.7 (many thanks to @SpenceKonde, it's great!!) with Arduino 1.8.19; working with an AVR128DB28 and flashing via SerialUSB UPDI (CH340-based).

I'm looking into Power Down sleep mode.

With AtMega 328p, I was able to achieve 200 nA with Power Down mode, WDT disabled, wake up by pin interrupt.

With AVR128DB, the best I managed is 1100 nA with Power Down mode, all pins to INPUT_PULLUP, BOD disabled.
What's worse, the chip won't wake up from pin level change.

So I have two questions, if I may:

1) Why the higher power draw in Power Down mode? I also tried to disable WDT (WDT.CTRLA = 0) to no effect; maybe the register is locked by FUSE.WDTCFG so I can't change it at runtime? Don't know how to modify fuses

2) Is there a working code example for wake up (from Power Down) after a pin change interrupt? I am obviously doing something wrong (just copied over my code from 328p, where it works)

Thanks, regards
Fernando

**Comment by XFer012** [comment]
Figured I may post my code as well :D

```
#include <avr/sleep.h>    
#include <avr/power.h>

#define ACTIVITY_MS 20000U

#define LED 7     // PA7 (upper-left pin)
#define SWITCH 6  // PA6 (upper-right pin) use this for edge change interrupt which wakes up the AVR, shorting it to GND
#define N_PINS 23
#define BLINK_DURATION_MS 500

// Globals
uint32_t gElapsed_ms = 0;


/////////
// ISR
/////////

void switchISR()
{	
  // Nothing to do
}

void setupPins()
{
  uint8_t i;

  for (i = 0; i < N_PINS; i++)
  {
    pinMode (i, INPUT_PULLUP);
  }
  pinMode (LED, OUTPUT);
}

void startSleep ()
{
  set_sleep_mode(SLEEP_MODE_PWR_DOWN);
  power_all_disable();        // power off ADC, Timer 0 and 1, serial interface
  // WDT.CTRLA = 0;           // Disable WDT (no improvements)
  // bitClear(RTC.CTRLA, 0);  // Disable RTC (worse, +20% power draw)
  // bitClear(RTC.PITCTRLA, 0);  // Disable Periodic Interrupt generator
  sleep_enable();
  sleep_cpu();				  // Enter sleep mode

  // Wakes up here
  
  sleep_disable();   
  power_all_enable();		// power everything back on
}  


void setupPinChangeInterrupt(const uint8_t switchPin)
{
  pinMode (switchPin, INPUT_PULLUP); // Momentary pushbutton will push it to GND
  attachInterrupt(digitalPinToInterrupt(switchPin), switchISR, CHANGE);
}


///////////
// SETUP
///////////

void setup ()
{
  setupPins();
}  


//////////
// LOOP
//////////

void loop ()
{
  // We just woke up: signal is 1 long blink
  digitalWrite (LED, HIGH);
  delay (2 * BLINK_DURATION_MS);
  digitalWrite (LED, LOW);
  
  // Some activity
  gElapsed_ms = millis();
  while (millis() - gElapsed_ms < ACTIVITY_MS)
  {
    delay(1);
  }

  // Signal we are going to sleep: 3 blinks
  digitalWrite (LED, HIGH);
  delay (BLINK_DURATION_MS);
  digitalWrite (LED, LOW);
  delay (BLINK_DURATION_MS);
  digitalWrite (LED, HIGH);
  delay (BLINK_DURATION_MS);
  digitalWrite (LED, LOW);
  delay (BLINK_DURATION_MS);
  digitalWrite (LED, HIGH);
  delay (BLINK_DURATION_MS);
  digitalWrite (LED, LOW);
  
  startSleep();
}  

```

**Comment by SpenceKonde** [comment]
Well there's yo u r first problem. Power_all_disable does nothing here. It
only works on classic avrs, they haven't updated that library yet (and I
wouldn't hold my breath for updates to those)


____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by FerGT50** [reply]
Ok, but even if power_all_disable does nothing, I'm still in Power Down mode, where according to the datasheet, all peripherals are disabled minus PIT and WDT.
In this condition, the advertised power draw should be between [EDITED according to latest datasheet] 700 nA at 25 C. I'm reading 1100 nA at 20 C.

Any suggestion, something I could investigate?

Will DxCore (and MegaTinyCore) offer some kind of sleep functions we can rely upon, sometime in the future?

**Comment by SpenceKonde** [comment]
Yes once I give the ATTinyCore users their 2.0.0 the stupid sleep library is next on the list.

Oh, do you have MVIIO enabled in the fuses? That takes an additional 500nA so it woulod come out just about exactly what you see?

**Reply by FerGT50** [reply]
MVIIO disabled, first thing I tried... :(

Any idea about the "does-not-wake-on-pin-change" problem? Did I miss something in my code? Should I somehow "enable wake up on pin change"?

**Comment by fsievers** [comment]
Did you read the errata for the AVR128? Maybe it's something like this:

**PD0_ Input Buffer is Floating**
On 28- and 32-pin package parts, the PD0 input buffer is floating. Because the default direction setting for PD0 is as an input pin, this may cause unexpected current consumption.
**Work Around**
Disable the PD0 input (ISC in PORTD.PIN0CTRL) or configure the pin as an output (bit 0 in PORTD.DIR).

There are also some other erratas that mentions increased power consumption.

https://ww1.microchip.com/downloads/en/DeviceDoc/AVR128DB28-32-48-64-SilConErrataClarif-DS80000915B.pdf

Edit: Update link to correct erratas revision

**Reply by XFer012** [reply]
There is no PD0 on AVR128DB28.
The errata about increased power consumption does not apply here (Vdd is 3.3V), nor the one about OpAmp increased draw (fixed in A5 and newer versions), but thanks for your contribution.

**Comment by SpenceKonde** [comment]
The issue with pd0 mentioned is specific to those parts that don't have the
the pin - they do have it internally it isn't connected to a pin that comes
out of the part, but can Stull result in increased power consumption just
like any other pin. Thats the bug. But I think DxCore sets it up as either
output or input pull up to prevent that...

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by XFer012** [reply]
Ah, sorry then, I did not catch the part about the "internal" PD0.
I will try to set it explicitely and measure again.

**Comment by XFer012** [comment]
Tried with pinMode (PIN_PD0, OUTPUT) , but still measuring 1100 nA.

Any chance PERIOD in FUSE.WDTCFG is set to a value > 0 by the core?
This would prevent the WDT to being disabled at runtime by WDT.CTRLA
